tls/s2n_client_cert_verify.c

```diff
@@ -33,27 +33,32 @@ int s2n_client_cert_verify_recv(struct s2n_connection *conn)
 
     struct s2n_stuffer *in = &conn->handshake.io;
 
-    POSIX_GUARD_RESULT(s2n_signature_algorithm_recv(conn, in));
-    const struct s2n_signature_scheme *chosen_sig_scheme = conn->handshake_params.client_cert_sig_scheme;
-    POSIX_ENSURE_REF(chosen_sig_scheme);
-
-    uint16_t signature_size = 0;
-    struct s2n_blob signature = { 0 };
-    POSIX_GUARD(s2n_stuffer_read_uint16(in, &signature_size));
-    signature.size = signature_size;
-    signature.data = s2n_stuffer_raw_read(in, signature.size);
-    POSIX_ENSURE_REF(signature.data);
-
-    /* Use a copy of the hash state since the verify digest computation may modify the running hash state we need later. */
-    struct s2n_hash_state *hash_state = &hashes->hash_workspace;
-    POSIX_GUARD_RESULT(s2n_handshake_copy_hash_state(conn, chosen_sig_scheme->hash_alg, hash_state));
-
-    /* Verify the signature */
-    POSIX_GUARD(s2n_pkey_verify(&conn->handshake_params.client_public_key, chosen_sig_scheme->sig_alg, hash_state, &signature));
+    if (conn->async_op.async_state == S2N_ASYNC_NOT_INVOKED) {
+        POSIX_GUARD_RESULT(s2n_signature_algorithm_recv(conn, in));
+        const struct s2n_signature_scheme *chosen_sig_scheme = conn->handshake_params.client_cert_sig_scheme;
+        POSIX_ENSURE_REF(chosen_sig_scheme);
+
+        uint16_t signature_size = 0;
+        struct s2n_blob signature = { 0 };
+        POSIX_GUARD(s2n_stuffer_read_uint16(in, &signature_size));
+        signature.size = signature_size;
+        signature.data = s2n_stuffer_raw_read(in, signature.size);
+        POSIX_ENSURE_REF(signature.data);
+
+        /* Use a copy of the hash state since the verify digest computation may modify the running hash state we need later. */
+        struct s2n_hash_state *hash_state = &hashes->hash_workspace;
+        POSIX_GUARD_RESULT(s2n_handshake_copy_hash_state(conn, chosen_sig_scheme->hash_alg, hash_state));
+
+        /* Verify the signature */
+        POSIX_GUARD(s2n_async_pkey_verify(conn, chosen_sig_scheme->sig_alg, hash_state, &signature));
+    } else if (conn->async_op.async_state == S2N_ASYNC_INVOKED) {
+        POSIX_BAIL(S2N_ERR_ASYNC_BLOCKED);
+    }
```

Comment:
Async pkey handled this with I think a macro? I wonder if that'd be clearer here. Like, I'm not sure every method that uses the async op should have to know how to handle all the possible values of async_state. I'm imagining it looking like:
```
S2N_ASYNC_OFFLOAD(conn,
    POSIX_GUARD_RESULT(s2n_signature_algorithm_recv(conn, in));
    const struct s2n_signature_scheme *chosen_sig_scheme = conn->handshake_params.client_cert_sig_scheme;
    POSIX_ENSURE_REF(chosen_sig_scheme);
    ....
    /* Verify the signature */
    POSIX_GUARD(s2n_async_pkey_verify(conn, chosen_sig_scheme->sig_alg, hash_state, &signature));
)
```

Here's an example of the kind of macro I'm talking about: https://godbolt.org/z/73PGxd36b

Reply:
You might even be able to work s2n_async_op_reset in there, if you call it immediately after the if/else instead of after the additional method logic. If you need the macro in both posix/result methods, you could make "guard_type" an argument and construct your guards like `guard_type##GUARD_RESULT`.